solutions/advanced/izho-12-xor.mdx

```diff
@@ -0,0 +1,152 @@
+
+We also check if $p_l \oplus p_r > x$ holds at this specific bit. If it does, then this means that the xor value is already greater and the less significant bits are irrelevant. So, this is an answer and we take the smallest index in the subtree rooted at this node as all the values in the subtree satisfy the original inequality.
+
+To recap, we will iterate through the array, adding the values into the trie as we go along. At every index, we find the smallest index satisfying the condition and are trivially able to find the longest interval satisfying the original condition.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log10^9)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+#define nl '\n'
+
+const int INF = 1e9 + 10;
+
+struct Trie {
+	int smallest_index;  // smallest index of a value in this sub-trie
+	Trie *children[2];   // (either 0 or 1)
+	Trie() {
+		smallest_index = INF;
+		children[0] = children[1] = NULL;
+	}
+
+	void insert_value(int value, int index, int bit = 30) {
+		// inserts x into the trie
+		smallest_index =
+		    min(smallest_index, index);  // update smallest index in sub-trie
+
+		if (bit == -1) return;
+		int value_bit = (value >> bit) & 1;
+
+		// create node if needed
+		if (!children[value_bit]) children[value_bit] = new Trie();
+
+		// recurse from child node
+		children[value_bit]->insert_value(value, index, bit - 1);
+	}
+
+	int find_smallest_index(int x, int y, int bit = 30) {
+		// returns smallest index such that P xor x >= y, where P is a element
+		// in the trie
+
+		if (bit == -1) return INF;
+		int current_answer = INF;
+
+		// get bit-th bits of x and y
+		int x_bit = (x >> bit) & 1, y_bit = (y >> bit) & 1;
+
+		for (int p_bit = 0; p_bit <= 1; p_bit++) {
+			// not possible (no values like this)
+			if (!children[p_bit]) continue;
+
+			if ((x_bit ^ p_bit) == y_bit) {
+				// possible answer, continue traversing
+				if (bit == 0) {
+					// reached equality, therefore is an answer
+					current_answer =
+					    min(current_answer, children[p_bit]->smallest_index);
+				} else {
+					// continue traversing
+					current_answer = min(
+					    current_answer,
+					    children[p_bit]->find_smallest_index(x, y, bit - 1));
+				}
+			}
+
+			if ((x_bit ^ p_bit) > y_bit) {
+				// already have P xor x > y, hence all other bits don't matter
+
+				// take smallest index in the sub-trie
+				current_answer =
+				    min(current_answer, children[p_bit]->smallest_index);
+
+				p_bit ^= 1;  // continue down other possibility
+			}
+		}
+
+		return current_answer;
+	}
+};
+
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
```

Comment:
If the code passes without fast IO, remove it. If it doesn't mention it within the editorial.

Reply:
damn i didn't know you were like this dustin

Reply:
Fixed